packages/instrumentation-amqplib/src/amqplib.ts

```diff
@@ -316,64 +329,164 @@ export class AmqplibInstrumentation extends InstrumentationBase<typeof amqp> {
+                            diag.error('amqplib instrumentation: publishConfirmHook error', e);
                         }
                     },
                     true
                 );
             }
 
-            // calling original publish function is only storing the message in queue.
-            // it does not send it and waits for an ack, so the span duration is expected to be very short.
+            const patchedOnConfirm = function (err: any, ok: amqp.Replies.Empty) {
+                // should we wrap with context and end span after this callback or end span right away?
+                context.with(trace.setSpan(context.active(), span), () => {
+                    callback?.call(this, err, ok);
+                });
+
+                if (self._config.publishConfirmEndHook) {
+                    safeExecuteInTheMiddle(
+                        () => self._config.publishConfirmEndHook(span, { exchange, routingKey, content, options }, err),
+                        (e) => {
+                            if (e) {
+                                diag.error('amqplib instrumentation: publishConfirmEndHook error', e);
+                            }
+                        },
+                        true
+                    );
+                }
+
+                if (err) {
+                    span.setStatus({
+                        code: SpanStatusCode.ERROR,
+                        message: "message confirmation has been nack'ed",
+                    });
+                } else {
+                    span.setStatus({ code: SpanStatusCode.OK });
+                }
+                span.end();
+            };
+
+            // calling confirm channel publish function is storing the message in queue and registering the callback for broker confirm.
+            // span ends in the patched callback.
             const argumentsCopy = [...arguments];
             argumentsCopy[3] = modifiedOptions;
-            const originalRes = original.apply(this, argumentsCopy);
-            if (!originalRes) {
-                span.setStatus({
-                    code: SpanStatusCode.ERROR,
-                    message: 'write buffer is full, message should be published again after drain event',
-                });
+            argumentsCopy[4] = patchedOnConfirm;
+            return original.apply(this, argumentsCopy);
+        };
+    }
+
+    protected getPublishPatch(
+        moduleVersion: string,
+        original: (exchange: string, routingKey: string, content: Buffer, options?: amqp.Options.Publish) => boolean
+    ) {
+        const self = this;
+        return function publish(
+            exchange: string,
+            routingKey: string,
+            content: Buffer,
+            options?: amqp.Options.Publish
+        ): boolean {
+            const isConfirmChannel = !!this.waitForConfirms; // duck typing, can we reliably check for CC in a cleaner way?
```

Comment:
A possible alternative would be to set some key on the context when starting the span on the CC, and then check here if this key exists here. something similar to `supressInstrumentation`, but specific to amqplib instrumentation.

Reply:
I went this way, not sure my propagation of context is proper here (I need to pass down mutated context to normal publish from CC publish) but it seems to work. Maybe style could be improved